### PR TITLE
Limit draw updates

### PR DESCRIPTION
**Pull request by graefjk**
This change limits the between individual calls of `drawEntities()`. This way  `drawEntities()` is not called every time a new Entity is created. This can reduce Lag when multible entities are added to the PlayingField at the same time. The time between `drawEntities()` calls can be changed with the `setTimeBetweenDraws` method which accepts the time in milliseconds. The time is initially set as 32ms between draw calls, which equates to a maximum of 31.75 Draws/Second. This implements #194.

### CODE REVIEW

File: ICGE-Simulation/src/main/java/de/unistuttgart/informatik/fius/icge/simulation/internal/playfield/StandardPlayfield.java [hunk 1]
```diff
@@ -54,6 +54,37 @@ public class StandardPlayfield implements Playfield {
     
     private Consumer<List<Drawable>> drawablesChangedListener;
     
+    private long             lastEntityDraw            = 0;
+    private volatile boolean delayedEntityDraw         = false;
```
Comment (buehlefs) [inline ICGE-Simulation/src/main/java/de/unistuttgart/informatik/fius/icge/simulation/internal/playfield/StandardPlayfield.java:58]:
```suggestion
    private volatile boolean awaitingEntityDraw        = false;
```
With this name, the `if` in `drawEntities` is easier to understand

**Review by buehlefs — COMMENTED** [review]
Looks good on first glance. The alternative to implementing the update batching/debouncing in the simulation is to implement it in the playfield drawer <https://github.com/FIUS/ICGE2/blob/49d372f95d025f954fada6a8dbabe216077d8b31/ICGE-Ui/src/main/java/de/unistuttgart/informatik/fius/icge/ui/internal/SwingPlayfieldDrawer.java#L192>
There you could use swing timers to trigger draw updates (this would be a bit nicer than the manual thread spawning and joining here). But I don't know where the biggest performance bottlenecks actually are.

**Comment by graefjk** [comment]
I chose to implement this here instead of in the playfield drawer since `drawEntities` executes `this.getAllEntities()` every time it is called and `this.getAllEntities()` takes longer the more entities there are.

File: ICGE-Simulation/src/main/java/de/unistuttgart/informatik/fius/icge/simulation/internal/playfield/StandardPlayfield.java [hunk 2]
```diff
@@ -83,24 +97,33 @@ public Simulation getSimulation() {
     }
     
     /**
-     * Converts all entities to drawables and sends them to the playfield drawer.
+     * Converts all entities to drawables and sends them to the playfield drawer every 32ms.
```
Comment (buehlefs) [inline ICGE-Simulation/src/main/java/de/unistuttgart/informatik/fius/icge/simulation/internal/playfield/StandardPlayfield.java:100]:
```suggestion
     * Queues a draw update to the playfield drawer. Draw updates are sent to the playfield drawer automatically every 32ms.
```

**Review by buehlefs — APPROVED** [review]
Looks good now. (But I did not run the code myself, so merge at your own risk.)

**Comment by graefjk** [comment]
> Looks good now. (But I did not run the code myself, so merge at your own risk.)

I cannot merge any pull request as i am not a maintainer of this project

**Comment by buehlefs** [comment]
It looks like I don't have the rights to make you a project maintainer. You should ensure that you have at least 1-2 active maintainers with the correct access rights (@neumantm should be able to help with that). If you want me to merge just leave me a thumbs up. (but I may not be available for emergency merges on a short notice)

**Pull request merged**